packages/ember-glimmer/lib/utils/custom-component-manager.ts

```diff
@@ -1,45 +1,27 @@
-import { assert } from '@ember/debug';
-import { Owner } from 'ember-owner';
-import { symbol } from 'ember-utils';
-
-import CustomComponentManager, { CustomComponentState } from '../component-managers/custom';
-
 import { GLIMMER_CUSTOM_COMPONENT_MANAGER } from '@ember/canary-features';
 
-export const COMPONENT_MANAGER = symbol('COMPONENT_MANAGER');
+const getPrototypeOf = Object.getPrototypeOf;
+const MANAGERS: WeakMap<any, string> = new WeakMap();
 
-export function componentManager(obj: any, managerId: String) {
-  if ('reopenClass' in obj) {
-    return obj.reopenClass({
-      [COMPONENT_MANAGER]: managerId,
-    });
-  }
+export function setComponentManager(managerId: string, obj: any) {
```

Comment:
The `set` language in the RFC supposedly give us the license to mutate the class. Would we want to do that instead of a manual prototype chain walk?

Reply:
The prototype walk is actually quite fast, and if we move this to setting on the object we break any objects that `Object.seal` or `Object.freeze` themselves (which is actually a good practice, and somewhat common).

Imagine this implementation for "template only components":

```js
owner.register('component-manager:pseudo-template-only', EmberObject.extend({
  capabilities: capabilities('3.4'),

  createComponent() {
    return null;
  },

  updateComponent() {},

  getContext() {
    return null;
  },
});

const BASE_COMPONENT = Object.freeze({});

setComponentManager('pseudo-template-only, BASE_COMPONENT);
```

If we were to mutate the object, we loose this...

Reply:
ah yes I remember discussing this